File: core/src/main/java/javabot/operations/ListAdminCommands.java
package javabot.operations;

import com.antwerkz.sofia.Sofia;
import javabot.Message;
import javabot.commands.AdminCommand;

import java.util.ArrayList;
import java.util.List;
import java.util.ServiceLoader;
import java.util.stream.Collectors;

public class ListAdminCommands extends AdminCommand {
    @Override
    public void execute(Message event) {
        List<String> list = listCommands()
                                .stream()
                                .map(AdminCommand::getCommandName)
                                .collect(Collectors.toList());
        getBot().postMessage(event.getChannel(), event.getUser(),
                             Sofia.adminKnownCommands(event.getUser().getNick(), String.join(", ", list)),
                             event.isTell());
    }

    public List<AdminCommand> listCommands() {
        final ServiceLoader<AdminCommand> loader = ServiceLoader.load(AdminCommand.class);
        final List<AdminCommand> list = new ArrayList<>();
        for (final AdminCommand command : loader) {
            list.add(command);
        }
        return list;
    }
}

File: src/main/kotlin/javabot/operations/ListAdminCommands.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import com.google.inject.Inject
import java.util.ServiceLoader
import javabot.Javabot
import javabot.Message
import javabot.commands.AdminCommand
import javabot.dao.AdminDao

class ListAdminCommands @Inject constructor(bot: Javabot, adminDao: AdminDao) :
    AdminCommand(bot, adminDao) {
    override fun execute(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        val list = listCommands().map({ it.getCommandName() })
        responses.add(
            Message(event, Sofia.adminKnownCommands(event.user.nick, list.joinToString(", ")))
        )
        return responses
    }

    fun listCommands(): List<AdminCommand> {
        val loader = ServiceLoader.load(AdminCommand::class.java)
        val list = arrayListOf<AdminCommand>()
        list.addAll(loader)
        return list
    }
}
